### PR TITLE
Api server makefile

### PR DESCRIPTION
**Pull request by z103cb**
## Why are these changes needed?

The value of these changes is to facilitate the local development of new features for the KubeRay API Server component.  It is a precursor to further work I and @blublinsky intend to do on integrating KubeRay API Server into the Codeflare stack.

## Related issue number

None.

## Checks

- [X] I've made sure the tests are passing. 
- Testing Strategy
   - [X] Unit tests
   - [X] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: apiserver/DEVELOPMENT.md [hunk 1]
```diff
@@ -7,8 +7,12 @@ This guide covers the purpose, requirements, and deployment of the Kuberay API S
 | Software | Version  |                                                            Link |
 | :------- | :------: | ------------------------------------------------------------: |
 | kubectl  | v1.18.3+ | [Download](https://kubernetes.io/docs/tasks/tools/install-kubectl/) |
-| Go       |  v1.13+  |                            [Download](https://golang.org/dl/) |
+| Go       |  v1.17   |                            [Download](https://golang.org/dl/) |
```
Comment (blublinsky) [inline apiserver/DEVELOPMENT.md:10]: Is it strictly 1.17 or 1.17+

**Review by blublinsky — APPROVED** [review]
LGTM

File: apiserver/DEVELOPMENT.md (hunk 1 above)
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:10]: For now strictly 1.17 as it the version defined in the `go.mod`. I have ran things with 1.18, no issues found.

**Comment by z103cb** [comment]
@Jeffwan, @scarlet25151 and @kevin85421 I would very much appreciate your feedback on this PR.

File: apiserver/DEVELOPMENT.md [hunk 2]
```diff
@@ -20,54 +24,83 @@ The backend service can be deployed locally or within a Kubernetes cluster. The
...
 
-This script will build and optionally push the image to the remote Docker Hub (hub.byted.org).
-
-#### Start Service
+#### Start Kubernetes Deployment
 
 ```bash
-kubectl apply -f deploy/
-```
+#To use the  helm charts
```
Comment (kevin85421) [inline apiserver/DEVELOPMENT.md:75]: Should we mention `make load-image` here?

Screen Shot 2023-08-08 at 6 04 53 PM

File: apiserver/Makefile [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	$(KUSTOMIZE) build deploy/base | kubectl delete -f -
+
+.PHONY: deploy
+deploy: ## Deploy via helm the kuberay api server to the K8s cluster specified in ~/.kube/config.
+	helm upgrade --install kuberay-apiserver ../helm-chart/kuberay-apiserver --wait \
+	--set image.tag=${IMG_TAG} --set image.pullPolicy=IfNotPresent
+
+.PHONY: undeploy
+undeploy: ## Undeploy via helm the kuberay api server to the K8s cluster specified in ~/.kube/config.
+	helm delete -n kuberay-apiserver kuberay-apiserver --wait
```
Comment (kevin85421) [inline apiserver/Makefile:103]: `make deploy` will install the chart in the `default` namespace.

```suggestion
	helm delete kuberay-apiserver --wait
```

File: apiserver/DEVELOPMENT.md [hunk 4]
```diff
@@ -20,54 +24,83 @@ The backend service can be deployed locally or within a Kubernetes cluster. The
...
+make deploy
 
-#### Access
+#To use the configuration from deploy/base
+make install
```
Comment (kevin85421) [inline apiserver/DEVELOPMENT.md:79]: Screen Shot 2023-08-08 at 6 14 34 PM

File: apiserver/hack/kind-cluster-config.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+- role: control-plane
+  image: kindest/node:v1.23.17@sha256:59c989ff8a517a93127d4a536e7014d28e235fb3529d9fba91b3951d461edfdb
+  kubeadmConfigPatches:
+    - |
+      kind: InitConfiguration
+      nodeRegistration:
+        kubeletExtraArgs:
+          node-labels: "ingress-ready=true"
+  extraPortMappings:
+  - containerPort: 30379
```
Comment (kevin85421) [inline apiserver/hack/kind-cluster-config.yaml:13]: Are there any use cases to forward GCS? I have never heard about any users forwarding the port.

File: helm-chart/kuberay-apiserver/templates/service.yaml [hunk 6]
```diff
@@ -5,7 +5,7 @@ metadata:
   annotations:
     prometheus.io/path: /metrics
     prometheus.io/scrape: "true"
-    prometheus.io/port: "8080"
+    prometheus.io/port: "8888"
```
Comment (kevin85421) [inline helm-chart/kuberay-apiserver/templates/service.yaml:8]: What's the reason behind this change?

File: apiserver/Makefile (hunk 3 above)
Comment (z103cb) [reply apiserver/Makefile:103]: Thanks, I will fix it.

File: apiserver/DEVELOPMENT.md [hunk 7]
```diff
@@ -20,54 +24,83 @@ The backend service can be deployed locally or within a Kubernetes cluster. The
...
+* `make clean-cluster` -- deletes the cluster created with the `cluster` target
+
+### Local Development
 
 #### Build
 
 ```bash
-go build -a -o raymgr cmd/main.go
+#To build the executable
+make build
```
[the remark this answers is not in the document]
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:38]: I can do that, but I think that it might be a bug. The `make build` should download the binary if it does not exist in the `../bin` directory.

File: helm-chart/kuberay-apiserver/templates/service.yaml (hunk 6 above)
Comment (z103cb) [reply helm-chart/kuberay-apiserver/templates/service.yaml:8]: To keep it alligned with the service definition in `deploy/base/appiserver.yaml` and the code behavior (metrics are exposed on port `8888`).

Here's the relevant bit in the `yaml` above:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: kuberay-apiserver
  annotations:
    prometheus.io/path: /metrics
    prometheus.io/scrape: "true"
    prometheus.io/port: "8888"
spec:

```

File: apiserver/deploy/base/apiserver.yaml [hunk 8]
```diff
@@ -18,6 +18,7 @@ spec:
       containers:
       - name: kuberay-apiserver
         image: kuberay/apiserver
+        imagePullPolicy: Always
```
[the remark this answers is not in the document]
Comment (z103cb) [reply apiserver/deploy/base/apiserver.yaml:21]: Oversight, will remove.

File: apiserver/DEVELOPMENT.md (hunk 2 above)
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:75]: Will fix, this is a bug.

File: apiserver/hack/kind-cluster-config.yaml (hunk 5 above)
Comment (z103cb) [reply apiserver/hack/kind-cluster-config.yaml:13]: @kevin85421 the kind configuration was lifted from the API server [official documentation] 
(https://ray-project.github.io/kuberay/components/apiserver/#swagger-support)

TL;DR here's the relevant bit
```yaml
    extraPortMappings:
      - containerPort: 30379
        hostPort: 6379
        listenAddress: "0.0.0.0"
        protocol: tcp
      - containerPort: 30265
        hostPort: 8265
        listenAddress: "0.0.0.0"
        protocol: tcp
      - containerPort: 30001
        hostPort: 10001
        listenAddress: "0.0.0.0"
        protocol: tcp
      - containerPort: 8000
        hostPort: 8000
        listenAddress: "0.0.0.0"
      - containerPort: 31888
        hostPort: 31888
        listenAddress: "0.0.0.0"
  - role: worker
  - role: worker
EOF
```

I can remove the mapping in this PR and update the documentation with a separate PR. How do you wish to proceed ?

File: apiserver/DEVELOPMENT.md (hunk 4 above)
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:79]: A bug, will fix it.

Comment (z103cb) [reply apiserver/DEVELOPMENT.md:94]: Fixed with [c4b4272](https://github.com/ray-project/kuberay/pull/1301/commits/c4b4272b4418b4248c6163556005b03b957f7e42)

File: apiserver/DEVELOPMENT.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:50]: [c4b4272](https://github.com/ray-project/kuberay/pull/1301/commits/c4b4272b4418b4248c6163556005b03b957f7e42)

File: apiserver/DEVELOPMENT.md [hunk 9]
```diff
@@ -20,54 +24,83 @@ The backend service can be deployed locally or within a Kubernetes cluster. The
...
+#To use the configuration from deploy/base
+make install
+```
 
-To obtain the port, run the following command:
+#### Stop Kubernetes Deployment
 
 ```bash
-NODE_PORT=$(kubectl get -o jsonpath="{.spec.ports[0].nodePort}" services backend-service -n ray-system)
+#To use the  helm charts
```
[the remark this answers is not in the document]
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:100]: Fixed with: [c4b4272](https://github.com/ray-project/kuberay/pull/1301/commits/c4b4272b4418b4248c6163556005b03b957f7e42)

File: apiserver/DEVELOPMENT.md (hunk 2 above)
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:90]: Assuming that this happening with a local kind cluster, I am guessing that the image is not loaded. What happens if you execute `make cluster load-image deploy`? 

I think that the easy fix is to add a dependency on the `load-image` target to load the image into the cluster. How do you wish to proceed ?

File: apiserver/hack/kind-cluster-config.yaml (hunk 5 above)
Comment (kevin85421) [reply apiserver/hack/kind-cluster-config.yaml:13]: > I can remove the mapping in this PR and update the documentation with a separate PR.

Great!

**Review by scarlet25151 — APPROVED** [review]
Overall LGTM.

File: apiserver/DEVELOPMENT.md (hunk 2 above)
Comment (z103cb) [reply apiserver/DEVELOPMENT.md:92]: Fixed with [c20bb27](https://github.com/ray-project/kuberay/pull/1301/commits/c20bb27ed42b7f1af41b2f79378fce6bcd868a3b)

File: apiserver/Makefile [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+	cd ../ray-operator && $(MAKE) docker-image -e IMG=kuberay/operator:$(OPERATOR_IMAGE_TAG)
+
+.PHONY: deploy-operator
+deploy-operator: ## Deploy operator via helm into the K8s cluster specified in ~/.kube/config.
+	helm upgrade --install raycluster ../helm-chart/kuberay-operator --wait \
+	--set image.tag=${OPERATOR_IMAGE_TAG} --set image.pullPolicy=IfNotPresent
+
+.PHONY: undeploy-operator
+undeploy-operator: ## Undeploy operator via helm from the K8s cluster specified in ~/.kube/config.
+	helm delete -n raycluster-operator raycluster --wait
```
[the remark this answers is not in the document]
Comment (z103cb) [reply apiserver/Makefile]: Fixed with [c20bb27](https://github.com/ray-project/kuberay/pull/1301/commits/c20bb27ed42b7f1af41b2f79378fce6bcd868a3b)

File: helm-chart/kuberay-apiserver/templates/service.yaml (hunk 6 above)
Comment (z103cb) [reply helm-chart/kuberay-apiserver/templates/service.yaml:8]: I have created this #1311 issue. We should probably discuss on that issue what is the best way to integrate.

File: apiserver/hack/kind-cluster-config.yaml (hunk 5 above)
Comment (z103cb) [reply apiserver/hack/kind-cluster-config.yaml]: Fixed in https://github.com/ray-project/kuberay/pull/1301/commits/c20bb27ed42b7f1af41b2f79378fce6bcd868a3b

File: apiserver/deploy/base/kustomization.yaml [hunk 11]
```diff
@@ -1,11 +1,21 @@
...
+    name: kuberay-apiserver
+    version: v1
 commonLabels:
-  app.kubernetes.io/name: kuberay
   app.kubernetes.io/component: kuberay-apiserver
+  app.kubernetes.io/name: kuberay
+images:
+- name: kuberay/apiserver
+  newName: kuberay/apiserver
+  newTag: latest
```
Comment (kevin85421) [inline apiserver/deploy/base/kustomization.yaml:21]: Is this YAML file intended only for development? If users are expected to deploy the KubeRay API Server using this file, we might need to use nightly as the tag, because `kuberay/apiserver:latest` isn't available on public image registries.

Comment (z103cb) [reply apiserver/deploy/base/kustomization.yaml:21]: I think it just for development, but that being said, let me try to see if I can accomodate both.

Comment (z103cb) [reply apiserver/deploy/base/kustomization.yaml]: I have added a local `kustomize.yaml` in [0c5eb2b](https://github.com/ray-project/kuberay/pull/1301/commits/0c5eb2b5ad7461a56ee83a78feb1881ff409b905). 

Let me know if you think I should add `install-nightly` and `uninstall-nightly` targets.

**Review by kevin85421 — APPROVED** [review]
LGTM. This is helpful! I will merge this PR after CI finishes.

File: apiserver/DEVELOPMENT.md [hunk 12]
```diff
@@ -4,70 +4,130 @@ This guide covers the purpose, requirements, and deployment of the Kuberay API S
...
 ## Requirements
 
-| Software | Version  |                                                            Link |
-| :------- | :------: | ------------------------------------------------------------: |
+| Software | Version  |                                                                Link |
+| :------- | :------: | ------------------------------------------------------------------: |
 | kubectl  | v1.18.3+ | [Download](https://kubernetes.io/docs/tasks/tools/install-kubectl/) |
-| Go       |  v1.13+  |                            [Download](https://golang.org/dl/) |
-| Docker   |  19.03+  |                  [Download](https://docs.docker.com/install/) |
+| Go       |  v1.17   |                                  [Download](https://golang.org/dl/) |
```
Comment (tedhtchang) [inline apiserver/DEVELOPMENT.md:10]: @blublinsky The go version does not match the minimal support go version in the [go.mod ](https://github.com/ray-project/kuberay/blob/2b8947ccec2eccf528b0e1d17c72241d441e9c54/apiserver/go.mod#L3)file.

Comment (tedhtchang) [reply apiserver/DEVELOPMENT.md:10]: @blublinsky No need to do anything. @kevin85421 created the [issue](https://github.com/ray-project/kuberay/issues/1329) to update the go version.
